### PR TITLE
Create rush-stack-compiler projects for several versions of TypeScript.

### PR DESCRIPTION
**Pull request by iclanton**

### CODE REVIEW

File: apps/api-extractor/src/ast/AstNamespace.ts [hunk 1]
```diff
@@ -105,13 +105,10 @@ export class AstNamespace extends AstModule {
       // the JSDoc comment Node can be found.
       // If there is no parent or grandparent of this VariableDeclaration then
       // we do not know how to obtain the JSDoc comment.
```
Comment (pgonzal) [inline apps/api-extractor/src/ast/AstNamespace.ts]: Damn...  This feature was broken by [https://github.com/Microsoft/web-build-tools/commit/4eca6649d9dd3cee6e8a2998400c201702f5e107](https://github.com/Microsoft/web-build-tools/commit/4eca6649d9dd3cee6e8a2998400c201702f5e107)

This entire file is deleted in AE7, so I guess it's not worth fixing

**Comment by pgonzal** [comment]
> 

Let's not do this.

If our standard always-enabled test runner fails for a project that doesn't have any unit tests, then our test runner is designed wrong.

---
Refers to: build-tests/rush-stack-compiler-2.4-library-test/src/TestClass.ts:6 in 78ea4d2. [](commit_id = 78ea4d2403511b946f40f91def71c443becc1fa8, deletion_comment = False)

File: build-tests/rush-stack-compiler-2.4-library-test/package.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "rush-stack-compiler-2.4-library-test",
```
Comment (pgonzal) [inline build-tests/rush-stack-compiler-2.4-library-test/package.json]: >ush-stack-compiler-2.4-library-test" [](start = 12, length = 36)

This seems like a lot of copy+pasting.  Is there any easy way to create a single project that tests all the compiler versions?

**Comment by pgonzal** [comment]
> }

By design, the RSC projects that are maintained in our master branch should all have identical API's, right?

I wonder if there's some way we could point them at the same .api.ts file and enforce that

---
Refers to: common/reviews/api/rush-stack-compiler-2.4.api.ts:6 in 78ea4d2. [](commit_id = 78ea4d2403511b946f40f91def71c443becc1fa8, deletion_comment = False)

File: core-build/gulp-core-build-typescript/src/ApiExtractorTask.ts [hunk 3]
```diff
@@ -6,7 +6,7 @@ import {
   JsonFile,
   FileSystem
 } from '@microsoft/node-core-library';
-import { ApiExtractorRunner } from '@microsoft/rush-stack-compiler';
+import { ApiExtractorRunner as TApiExtractorRunner } from '@microsoft/rush-stack-compiler';
```
Comment (pgonzal) [inline core-build/gulp-core-build-typescript/src/ApiExtractorTask.ts]: >TApiExtractorRunner [](start = 31, length = 19)

Why is this renamed?

File: stack/rush-stack-compiler-2.9/src/TslintRunner.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      }
+    );
+  }
+
+  public invoke(): Promise<void> {
+    const args: string[] = [
+      '--format', 'json',
+      '--project', this._standardBuildFolders.projectFolderPath
+    ];
+
```
Comment (pgonzal) [inline stack/rush-stack-compiler-2.9/src/TslintRunner.ts]: Hrmmm.... this is a LOT of copy+pasting.  How are we supposed to keep these files in sync?

File: build-tests/rush-stack-compiler-2.4-library-test/package.json (hunk 2 above)
Comment (iclanton) [reply build-tests/rush-stack-compiler-2.4-library-test/package.json]: Discussed offline and fixed.

---
In reply to: [234095426](https://github.com/Microsoft/web-build-tools/pull/947#discussion_r234095426) [](ancestors = 234095426)

File: core-build/gulp-core-build-typescript/src/ApiExtractorTask.ts (hunk 3 above)
Comment (iclanton) [reply core-build/gulp-core-build-typescript/src/ApiExtractorTask.ts]: I want to specifically designate that this should only be used as a type.

---
In reply to: [234095868](https://github.com/Microsoft/web-build-tools/pull/947#discussion_r234095868) [](ancestors = 234095868)

File: apps/api-documenter/package.json [hunk 5]
```diff
@@ -22,7 +22,7 @@
     "js-yaml": "~3.9.1"
   },
   "devDependencies": {
-    "@microsoft/rush-stack-compiler": "0.5.4",
+    "@microsoft/rush-stack-compiler-3.0": "0.0.0",
```
Comment (pgonzal) [inline apps/api-documenter/package.json]: > "0.0.0", [](start = 41, length = 9)

0.0.0?

Comment (iclanton) [reply apps/api-documenter/package.json]: It'll get bumped to 0.1.0 when it's published.

---
In reply to: [236908556](https://github.com/Microsoft/web-build-tools/pull/947#discussion_r236908556) [](ancestors = 236908556)

**Comment by pgonzal** [comment]
> 

Why do we need to duplicate this file?

Maybe package.json's "main" entry should point to "./lib/shared/index.js" until we encounter an actual need to break away

---
Refers to: stack/rush-stack-compiler-2.7/src/index.ts:22 in 92c394f. [](commit_id = 92c394fa127d7356a5ab90ed3d36895f1c55e485, deletion_comment = False)

File: stack/rush-stack-compiler-2.7/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    "outDir": "lib"
+  },
+  "include": [
+    "src/**/*.ts",
+    "src/**/*.tsx"
+  ],
+  "exclude": [
+    "node_modules",
+    "lib"
+  ]
```
Comment (pgonzal) [inline stack/rush-stack-compiler-2.7/tsconfig.json]: Delete
